src/mono/wasi/runtime/main.c

```diff
@@ -10,7 +11,31 @@ const char* dotnet_wasi_getentrypointassemblyname();
 WASI_AFTER_RUNTIME_LOADED_DECLARATIONS
 #endif
 
-int main(int argc, char * argv[]) {
+#ifdef WASM_LIBRARY_MODE
+// _initialize is a function generated by the WASI SDK libc that calls the LLVM synthesized __wasm_call_ctors function for reactor components:
+// https://github.com/WebAssembly/wasi-libc/blob/9f51a7102085ec6a6ced5778f0864c9af9f50000/libc-bottom-half/crt/crt1-reactor.c#L7-L27
+// We define and call it for WASM_LIBRARY_MODE and TARGET_WASI to call all the global c++ static constructors.  This ensures the runtime is initialized
+// when calling into WebAssembly Component Model components.
+extern void _initialize();
+
+// CustomNativeMain programs are built using the same libbootstrapperdll as WASM_LIBRARY_MODE but wasi-libc will not provide an _initialize implementation,
+// so create a dummy one here and make it weak to allow wasi-libc to provide the real implementation for WASI reactor components.
+__attribute__((weak)) void _initialize()
```

Comment:
From what I understood about `_initialize`, this is an important bit https://github.com/dotnet/runtimelab/blob/feature/NativeAOT-LLVM/src/coreclr/nativeaot/Bootstrap/main.cpp#L194. We don't support custom unmanaged main, so maybe this is not needed on Mono?

@dicej @SingleAccretion Please, do you have more details about `_initialize`?

Reply:
`_initialize` is part of the WASI reactor ABI. It is similar to `DllMain` on Windows. One responsibility of `_initialize` is to run global (C++) constructors (`__attribute__ ((constructor))` methods).

The code being copied here is a workaround for the fact that global constructors are/were broken with components (https://github.com/dotnet/runtimelab/issues/2359).

The workaround as copied here is not correct, as you pointed out. The `_initialize` call needs to be guarded on `if (<global constructors have not been called yet>)`.

Reply:
Yeah, `_initialize` should be called at most once, per https://github.com/WebAssembly/WASI/blob/main/legacy/application-abi.md and https://github.com/WebAssembly/tool-conventions/pull/203

Reply:
Thanks, updated to call `_initialize` only once.